### DISCUSSION TITLE
Is there a way to have auto imports in zed?

### DISCUSSION THREAD

**Discussion by abusada** [zed-industries/zed#13522] [Feature Requests]
Loving zed, but I have come to heavily depend on auto import feature provided by https://github.com/NuclleaR/vscode-extension-auto-import in vscode, Is there currently a way to achieve the same result in zed?

**Comment by omarjatoi** [comment]
I think this is possible using either [`code_actions`](https://github.com/zed-industries/zed/blob/main/docs/src/configuring-zed.md#formatter) or [`code_actions_on_format`](https://github.com/zed-industries/zed/blob/main/docs/src/configuring-zed.md#code-actions-on-format) and the [`source.addMissingImports.ts`](https://github.com/typescript-language-server/typescript-language-server?tab=readme-ov-file#code-actions-on-save) `code_action` from the [TypeScript Language Server](https://github.com/typescript-language-server/typescript-language-server). Side note, this is language specific.

**Reply by anurag-zuddl** [reply]
@omarjatoi Any idea how we can configure this? I was not successful in finding a solution.

**Reply by thataboy** [reply]
> @omarjatoi Any idea how we can configure this? I was not successful in finding a solution.

try reading here [configure lsp](https://zed.dev/docs/configuring-zed?highlight=lsp#lsp)

**Comment by colinbes** [comment]
Totally code specific but without this feature I find using Zed Is not worth it which is a huge pity as it has so much to offer. But hunting down import locations is way slow when up to me. Unless I’m missing something ….

**Comment by BatmanAoD** [comment]
VSCode provides importing as a "quick fix" code action, which is often preferable to importing on save, because it lets the user select the correct import from multiple potential sources.

**Reply by 88potatoes** [reply]
There is https://github.com/zed-industries/zed/discussions/13115#discussioncomment-9787758

**Comment by shamwela** [comment]
1. Open your `settings.json` file. Press `Command + ,`
2. Add this.
```
{
  "formatter": {
    "code_actions": {
      "source.addMissingImports.ts": true
    }
  },
 // Other settings here ...
}
```

**Reply by BatmanAoD** [reply]
Does this let you select from a list of imports, and does it work for other languages?

**Reply by colinbes** [reply]
> 1. Open your `settings.json` file. Press `Command + ,`
> 2. Add this.
> 
> ```
> {
>   "formatter": {
>     "code_actions": {
>       "source.addMissingImports.ts": true
>     }
>   },
>  // Other settings here ...
> }
> ```

What should this do? Add import as you type, when you save? I don't see it adding imports at all?

**Comment by marcoancona** [comment]
I am afraid the lack of auto-import (as a list to pick the correct one) is also a deal breaker for me

**Comment by thisloke** [comment]
I'm using Zed with zero custom configuration and i found that just pressing:
`Command + .
`![image](https://github.com/user-attachments/assets/37a82e11-6744-426c-900c-4d9cf490ab8d)

It works like a charm :D

**Reply by BatmanAoD** [reply]
Thank you! Not sure if this is new or if I we all just missed it.

**Reply by addy** [reply]
This is very hit or miss for me, sometimes I get the option, sometimes I don't (specific to TS repos only)

**Reply by yoandresaav** [reply]
worked like a charm!

**Reply by suryapandian** [reply]
no luck for me :(

**Comment by mishamsk** [comment]
Curious if there are any plans to add Python support as well?

**Reply by mishamsk** [reply]
TIL (replying to myself) - apparently it works partially. If you start typing a name and it pops up in LSP server suggestions, then completing the symbol will also auto-add import. However, if the entire name is already typed in, I couldn't find a way to trigger auto-import. Unlike JS/TS - there is no code action for it.

**Reply by neuromantic99** [reply]
`cmd + .` on a mac seems to work for me on python files now (similar to vscode)

**Comment by adamerose** [comment]
The distinction between the `Ctrl+Space` and `Ctrl+.` menus in VS Code and Zed is a little hazy to me, but VS Code shows auto-import in both while Zed only shows it in the latter - this should probably be fixed?
